samples/DevTestLabs/Modules/Library/Az.DevTestLabs2.psm1

```diff
@@ -0,0 +1,2636 @@
+    [ValidateNotNullOrEmpty()]
+    $Lab,
+
+    [parameter(Mandatory=$false,HelpMessage="Name of the vms to retrieve.  This parameter supports wildcards at the beginning and/or end of the string.")]
+    [ValidateNotNullOrEmpty()]
+    [string]
+    $Name = "*",
+
+    [parameter(Mandatory=$false,HelpMessage="Name of the vms to retrieve.  This parameter supports wildcards at the beginning and/or end of the string.")]
+    [ValidateSet('Starting', 'Running', 'Stopping', 'Stopped', 'Deallocating', 'Deallocated', 'Any')]
+    [string]
+    $Status = 'Any'
+  )
+
+  begin {. BeginPreamble}
+  process {
+    try {
+      foreach($l in $Lab) {
+        $ResourceGroupName = $l.ResourceGroupName
+        $LabName = $l.Name
+        Write-verbose "Retrieving $Name VMs for lab $LabName in $ResourceGroupName"
+        # Need to query client side to support wildcard at start of name as well (but this is bad as potentially many vms are involved)
+        # Also notice silently continue for errors to return empty set for composibility
+        # TODO: is there a cleaver way to make this less expensive? I.E. prequery without -ExpandProperties and then use the result to query again.
+        $vms = MyGetResourceVm -Name "$Name" -LabName $LabName -ResourceGroupName $ResourceGroupName
+        Write-verbose "Vms before status filter are $vms"
+        if($vms -and ($Status -ne 'Any')) {
+          return $vms | Where-Object { $Status -eq (Get-AzDtlVmStatus $_)}
+        } else {
+          return $vms
+        }
+      }
+    } catch {
+      Write-Error -ErrorRecord $_ -EA $callerEA
+    }
+  }
+  end {}
+}
+
+function Get-AzDtlVmStatus {
+  [CmdletBinding()]
+  param(
+    [parameter(Mandatory=$true,HelpMessage="VM to start. Noop if already started.", ValueFromPipeline=$true)]
+    [ValidateNotNullOrEmpty()]
+    $Vm
+  )
+
+  begin {. BeginPreamble}
+  process {
+    try {
+      foreach($v in $Vm) {
+
+        if(-not $v.Properties.lastKnownPowerState) {
+          return "Stopped"
+        } else {
+          return $v.Properties.lastKnownPowerState
+        }
+
+        # This is how to get it from the compute VM if the above turns out not to work
+<#         $computeVm = GetComputeVm($v)
+        $computeGroup = $computeVm.ResourceGroupName
+        $name = $computeVm.Name
+
+        $compVM = Get-AzureRmVM -ResourceGroupName $computeGroup -name $name -Status
+
+        $statuses = $compVM.Statuses
+        if(-not $statuses) {
+          throw "Can't get status for VM $name in $computeGroup"
+        }
+        return ($statuses | Where Code -Like 'PowerState/*')[0].DisplayStatus
+ #>
+      }
+    } catch {
+      Write-Error -ErrorRecord $_ -EA $callerEA
+    }
+  }
+  end {}
+}
+
+function Start-AzDtlVm {
+  [CmdletBinding()]
+  param(
+    [parameter(Mandatory=$true,HelpMessage="VM to start. Noop if already started.", ValueFromPipeline=$true)]
+    [ValidateNotNullOrEmpty()]
+    $Vm
+  )
+
+  begin {. BeginPreamble}
```

Comment:
I considered that, but it doesn't make things much better because of the  `. BeginPreamble` statement, that cannot be inside an utility function, so you end up replacing just one line of code with little benefit. I will send via email detials on the rationale for the errror management in the library